feat: 카카오 로그인

duol9 opened the pull request:
## 🌱 관련 이슈
- close #39 

## 📌 작업 내용 및 특이사항
- JWT 적용
- 시큐리티 적용
- 카카오 로그인 API 구현
- 토큰 재발급 API 구현

## 📝 참고사항
- 카카오 인증 코드를 얻기 위해 리다이렉트 uri가 추가되었습니다. 
- 토큰 재발급 API가 추가되었습니다. 
- 제가 맡았던 방과 회원 쪽은 시큐리티를 통해 회원정보를 가져오는 걸로 수정했습니다. 노트 쪽도 해보려 했으나 생각보다 로직이 추가될 것 같아 건들지는 못했습니다. 
- AuthMember로 Member 객체를 조회할 수 있는 MemberUtil.getMember 메서드를 추가했습니다. 

## 📚 기타
-


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - 카카오 OAuth 로그인(로그인 URL·콜백), 액세스·리프레시 토큰 발급·재발급, 로그아웃(토큰 블랙리스트 포함) 지원
  - JWT 생성·검증·필터·엔트리포인트·관련 오류 코드·인증 멤버(AuthMember) 추가
  - 카카오 연동 HTTP 클라이언트/DTO, 리프레시 토큰 엔티티·저장소·토큰 서비스, 캐시 기반 토큰 블랙리스트 및 캐시 설정 추가

- **Refactor**
  - 회원/방/알림/푸시 토큰 API를 인증된 사용자 컨텍스트 기반으로 전환 (MemberUtil 도입)

- **Chores**
  - 보안·JWT·캐시(Caffeine)·HTTP 클라이언트 의존성 추가
  - CI에 테스트·문제 리포트 업로드 단계 추가
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

duol9 committed:
chore: jwt 설정

duol9 committed:
feat: JWT, Security 적용

duol9 committed:
feat: AuthMember 작성

duol9 committed:
feat: @Auth 어노테이션 추가

- 기존 어노테이션 이름이 길어 짧게 쓰고자 생성

duol9 committed:
chore: RestClient 설정

duol9 committed:
feat: 카카오 로그인 통신 로직 구현

duol9 committed:
feat: RefreshToken 엔티티 작성

duol9 committed:
feat: Token 기본 Dto 작성

duol9 committed:
feat: RefreshTokenRepository 추가

duol9 committed:
feat: 토큰 생성 서비스 로직 구현

duol9 committed:
feat: Member 생성자 구현

duol9 committed:
feat: OauthId로 Member 조회하는 메서드 추가

duol9 committed:
feat: 카카오 로그인 서비스 로직 구현

duol9 committed:
feat: 카카오 로그인 API 추가

duol9 committed:
chore: oauth, jwt 프로파일 적용

- dev 프로파일에 cors url도 추가
- 환경변수 자동 매핑되도록 추가

duol9 committed:
feat: 토큰으로 memberId 조회하는 메서드 구현

duol9 committed:
feat: RefreshToken 업데이트 메서드 구현

duol9 committed:
feat: 토큰 재발급 서비스 로직 구현

- 추가 변경 사항 : generateToken -> generateTokens 로 메서드명 변경

duol9 committed:
feat: 기존 RefreshToken 값으로 RefreshToken 객체 조회하는 메서드 추가

duol9 committed:
feat: 토큰 재발급 API 추가

duol9 committed:
refactor: 토큰 발급 메서드명 변경

duol9 committed:
feat: AuthMember로 Member 객체 조회하는 메서드 구현

duol9 committed:
refactor: Member 정보 필요한 API들 수정

duol9 committed:
chore: cors 관련 jwt 프로파일로 이동

duol9 committed:
fix: 테스트 환경에서 SecurityConfig 무시

duol9 committed:
fix: 테스트 환경에서 무시

duol9 committed:
chore: CI 리포트 업로드 추가

duol9 committed:
test: 테스트용 Config 생성

duol9 committed:
test: TestConfig 적용하도록 설정

duol9 committed:
test: test yml 파일 생성

duol9 committed:
test: test yml에 필요한 값 추가

duol9 committed:
refactor: RefreshTokenService 클래스명 변경

- 범용적으로 쓰고자 TokenService로 변경

duol9 committed:
chore: 캐시 설정

duol9 committed:
feat: 회원 기준 토큰 삭제 메서드 추가

duol9 committed:
feat: 토큰 삭제 및 무효화 서비스 로직 구현

- addAccessTokenToBlacklist: accessToken을 캐싱
- isAccessTokenBlacklisted: 캐싱된 accessToken이 있는지 검사

duol9 committed:
feat: jwt 필터에서 무효화된 토큰 검증하는 로직 추가

duol9 committed:
feat: 로그아웃 API 추가

duol9 committed:
fix: 시큐리티 필터로 인해 swagger 접속 안되는 문제 해결

ohhamma commented on src/main/java/com/example/wini/global/config/SecurityConfig.java:
```diff
@@ -0,0 +1,82 @@
+    private final JwtAuthenticationFilter jwtAuthenticationFilter;
+
+    @Bean
+    @Order(0)
```
Bean에 순서를 매기는건 처음 보네요! 좋은 정보 알아갑니다 🙂

ohhamma commented on src/main/java/com/example/wini/global/security/JwtAuthenticationException.java:
JwtAuthenticationException을 따로 만들어주신 이유가 궁금합니다!

duol9 replied:
CustomException으로 처리해도 되긴 하지만 해당 오류는 비즈니스 오류라 각 오류가 발생한 계층(비즈니스 vs 보안 필터)에서 적절하게 처리되도록 설계하고자 했습니다. 
그리고 필터에서 일어나는 예외는 GlobalExceptionHandler에 도달하지 못해서 따로 처리도 해야하는 김에 이름으로 명확성을 두고 싶었습니다!

ohhamma commented on src/main/java/com/example/wini/domain/common/annotation/Auth.java:
커스텀 어노테이션 만들어주신 것도 정말 유용해보이네요! 잘 사용하겠습니다 🙂

ohhamma commented on src/main/java/com/example/wini/global/config/RestClientConfig.java:
비동기식 WebClient를 사용하지 않고 동기식 RestClient를 사용하신 이유가 궁금합니다!

duol9 replied:
해당 로그인 과정은 동기적으로 처리해야 하기도 하고 RestClient가 저에게 더 친숙하고 코드적으로도 가독성이 더 좋아서 선택했습니다!

ohhamma commented on src/main/java/com/example/wini/infra/kakao/client/KakaoClient.java:
```diff
@@ -0,0 +1,83 @@
+                .retrieve()
+                .onStatus(HttpStatusCode::is4xxClientError, (req, res) -> {
+                    throw new RuntimeException(
+                            "[KakaoClient] 카카오 사용자 정보 요청에 실패하였습니다. HTTP Status: " + res.getStatusCode());
```
여기서 CustomException이 아닌 RuntimeException을 던진 이유가 궁금합니다!

duol9 replied:
아 요건 전 코드 긁어오면서 수정을 못한거 같습니다... 허허 발견해주셔서 감사합니다!

ohhamma commented on src/main/java/com/example/wini/infra/kakao/dto/KakaoProfile.java:
저희 응답 dto 형식을 snake_case로 할지 camelCase로 할지 같이 이야기해보면 좋을 것 같습니다.

duol9 replied:
아 요건 음 카카오에서 전달하는 값들을 담기 위한 dto라서 어쩔수 없이 snake_case를 작성했습니다. 일치하지 않으면 받아오지 못하더라구요. 
서비스 Dto는 camelCase 로 쭉 가도 괜찮을거 같아요. 초기에 컨벤션 정할 때도 camelCase 하기로 하기도 했고요!

ohhamma replied:
아하 그렇군요! 확인했습니다 :)

ohhamma commented on src/main/java/com/example/wini/domain/member/repository/MemberRepository.java:
```diff
@@ -1,6 +1,10 @@
-public interface MemberRepository extends JpaRepository<Member, Long>, MemberCustomRepository {}
+public interface MemberRepository extends JpaRepository<Member, Long>, MemberCustomRepository {
+
+    Optional<Member> findByOauthId(String oauthId);
```
여기서 oauthProvider를 포함해서 사용자를 조회해오는게 더 좋을 것 같습니다.
다른 oauthProvider이어도 oauthId가 같은 경우에는 충돌이 일어나는 경우가 발생 위험이 있을 것 같아요.

duol9 replied:
음 넵! 구글, 카카오, 애플 모두 길이가 달라 딱히 신경쓰진 않았던 부분입니다. 반영하겟습니다.

ohhamma commented on src/main/java/com/example/wini/domain/auth/service/RefreshTokenService.java:
```diff
@@ -13,16 +19,51 @@
+        return updateAndGenerateTokens(refreshToken);
+    }
+
+    private String validateRefreshToken(String refreshTokenHeader) {
```
validate 함수는 주로 void를 반환하는 경우가 많더라고요!
대신 함수 안에서 조건을 검사하고 예외를 던지는 역할을 맡죠.

jwtProvider로부터 refreshToken을 가져오는 로직을 validate 밖으로 빼는건 어떨까요?

duol9 replied:
넵 반영하겟씁니다!

ohhamma commented on src/main/java/com/example/wini/domain/auth/service/RefreshTokenService.java:
```diff
@@ -13,16 +19,51 @@
+    private TokenResponse updateAndGenerateTokens(String refreshToken) {
+        Long userId = jwtProvider.getMemberId(refreshToken, REFRESH_TOKEN);
+
+        Member member = memberRepository.findById(userId).orElseThrow(() -> new CustomException(MEMBER_NOT_FOUND));
```
memberId가 아닌 userId를 사용하신 이유가 궁금합니다!

duol9 replied:
ㅎ.. 이것도 제가 수정을 다 했다 생각했는데 놓친것 같아요. 전 프로젝트에서 user를 사용햇엇습니다

ohhamma commented on src/main/java/com/example/wini/domain/common/util/MemberUtil.java:
```diff
@@ -0,0 +1,25 @@
+    private final MemberRepository memberRepository;
+
+    @Transactional(readOnly = true)
+    public Member getMember(AuthMember authMember) {
```
`getCurrentMember()`와 같이 조금 더 직접적인 함수명을 사용해도 좋을 것 같습니다 🙂

duol9 replied:
흠 하긴 일반 member를 가져오는 거로 혼동이 올 수도 있는 이름인것 같네요!

ohhamma commented on src/main/java/com/example/wini/domain/member/controller/MemberController.java:
하영님 혹시 모든 컨트롤러 파라미터로 AuthMember가 필수적으로 들어가야하나요?

duol9 replied:
로그인한 멤버 정보가 필요한 곳만 쓰면 됩니다! 그래서 단순 조회(템플릿, 상태 등) 말고는 거의 다 들어가야 할 것 같아요

ohhamma replied:
그렇군요 인증인가 때문에 모든 컨트롤러 함수에 파라미터가 늘어나는 것을 방지하기 위해
MemberUtil에 아래와 같은 방식으로 구현해서 호출하는건 어떤가요?

```java
	public Member getCurrentMember() {
		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
		Long memberId = getCurrentMemberId(authentication.getName());
		return memberRepository.findById(memberId)
			.orElseThrow(() -> new CustomException(MEMBER_NOT_FOUND);
	}
```

duol9 replied:
코드가 줄이기 위해서라는 점에선 공감합니다!
그런데 추후 테스트코드를 작성할 일이 있으면 SecurityContextHolder를 Mocking 하는게 조금 어렵진 않을까? 싶은 생각도 들어요 (해본 적이 없긴 합니다..!)

그럼 우선 제안해주신 방식을 적용하고 유지보수 방향 측면에선 문제가 생기면 그때 생각해 보는 걸로 할까요?

ohhamma replied:
좋습니다! 테스트코드 작성할때 다시 한번 이야기해봐도 좋을 것 같습니다.

ohhamma commented on src/main/java/com/example/wini/global/config/CacheConfig.java:
이부분은 redis 연결 대용으로 추가하신 부분일까요?!

duol9 replied:
그걸 의도한 건 아니에요!
스프링에서 기본적으로 제공해주는 캐시 기능에는 TTL 설정을 하지 못해서 기본 캐시 위에 입힐 라이브러리를 따로 추가하느라 생성했습니다!

ohhamma reviewed (commented):
인증인가 구현하느라 정말 수고많으셨습니다! 코드량이 어마어마하네요
작성해주신 부분 쭉 훑어보면서 궁금했던 점이나 개선이 필요한 점 등에 코멘트 달아두었습니다.

duol9 committed:
refactor: 카카오 유저 정보 조회 시 예외처리 컨벤션에 맞게 수정

duol9 committed:
refactor: 리프레시 토큰 유효성 검증 로직 개선

duol9 committed:
refactor: RefreshToken 재발급 로직을 RefreshTokenService로 분리

- reissueToken 메서드를 별도 서비스로 분리하여 재발급 로직의 책임 명확화
- TokenService는 토큰 생성, 갱신, 삭제 등 핵심 토큰 관리 로직만 담당
- 생성만 했던 기존 generateToken을 upsertTokens로 변경하여 업데이트(update) 혹은 생성(insert)을 담당도록 수정

duol9 committed:
refactor: 멤버ID 변수명 변경

duol9 committed:
refactor: 인증 회원 조회 메서드명 변경

duol9 committed:
feat: JpaAuditingConfig에 회원 정보 입력되도록 구현

duol9 committed:
feat: Swagger 문서에 Bearer 보안 스키마 적용

- 로그인 등 시큐리티 필터가 필요하지 않는 곳엔 @SecurityRequirements(value = {})를 적용해 보안 요구사항 무시

ohhamma commented on src/main/java/com/example/wini/domain/auth/service/AuthService.java:
```diff
@@ -37,6 +37,6 @@ public TokenResponse kakaoLogin(String authCode) {
         });
 
-        return tokenService.generateTokens(member);
+        return tokenService.upsertTokens(member);
```
`upsertTokens`는 어떤 함수인가요?

duol9 replied:
기존엔 토큰이 생성될 때마다 db에 save가 되었는데 현재 토큰과 멤버는 1대1 관계라 오류기 날 것 같았습니다. 
리펙토링 선택지가 아래 중 하나였는데 update를 하는게 효율적이라 판단했습니다. 
1. update 
2. delete 후 insert
 하지만 회원가입이라면 refreshToken이 insert 되는거고, 기존 회원이라면 update가 되어야 하는데 이걸 토큰이 생성될 때마다 분기점을 줘야 하는 부분이라 insert 또는 update를 할 수 있게 하는 메서드로 리팩토링 했습니다. 

참고로 upsert는 update + insert의 합성어라고 합니다! 자주 쓰이는 단어라고 하는데 저도 이번에 처음 알게 되었어요.

ohhamma replied:
오호 그렇군요 저도 upsert라는 표현이 낯설어서 여쭈어봤습니다. 알려주셔서 감사합니다!

ohhamma reviewed (commented):
수정해주신 부분 확인했습니다!
한가지 궁금한 점만 코멘트로 달아두었습니다.

duol9 committed:
refactor: oauth 정보로 회원 조회하는 메서드 수정

- oauthId 외에 provider도 일치해야 조회할 수 있도록 수정

duol9 committed:
refactor: 카카오 로그인api 리다이렉트 방식으로 변경

- 기존: /login/kakao 호출 시 /kakao/callback으로 리다이렉트 되는 주소 반환 (httpStatus 200)
- 현재: /login/kakao 호출 시 반환되는 주소를 통해 리다이렉트 (httpStatus 302)

duol9 committed:
refactor: 현재 로그인 회원 조회 로직을 SecurityContext 기반으로 수정

- memberId만 찾는 경우도 있어 메서드를 따로 분리하였습니다.

ohhamma reviewed (approved):
MemberUtil 수정해주신 부분까지 모두 확인했습니다! 수고많으셨습니다
